### PR TITLE
Add support for reason-language-server

### PR DESCRIPTION
**Pull request by deecewan**
I've read through `:h ale-dev`. I have *not* added linter tests. I'm not really even sure how they'd work, because you need extra binaries.

Let me know if/how I should add tests if they are required.

### CODE REVIEW

File: ale_linters/reason/reason_ls.vim [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    return ''
+endfunction
+
+call ale#linter#Define('reason', {
+\   'name': 'reason-language-server',
+\   'lsp': 'stdio',
+\   'executable':{buffer -> ale#Var(buffer, 'reason_ls_executable')},
+\   'command': '%e',
+\   'project_root': function('ale_linters#reason#reason_ls#FindProjectRoot'),
+\   'language': 'reason',
```
Comment (w0rp) [inline ale_linters/reason/reason_ls.vim]: The `language` option can be removed if the first argument to `Define` is the same string.

**Comment by w0rp** [comment]
You don't need extra binaries. Read through the help file and look existing tests in the `test` directories mentioned in the help file. The tests just check the string for the command that will be be run. For language server tests, you can test the project root detection function by adding some test directories with empty files. There are a few tests which do this you can look at for examples, such as the `test_rust_rls_callbacks.vader` file.

**Comment by stouf** [comment]
Any news on this? It would be very nice to have 😊 Please don't hesitate to call out for help if needed :+1:

**Comment by w0rp** [comment]
If there are any updates, you would already see them here. This needs tests.

**Comment by stouf** [comment]
> If there are any updates, you would already see them here. This needs tests.

Yeah, I was just trying to get back some attention on this. And also letting know @deecewan I can try to help if he needs anything.

**Comment by deecewan** [comment]
please, go ahead an add tests if you feel comfortable to do so. I don't have a pressing need for this. I was just playing with reason at the time and noticed the gap.

**Comment by stouf** [comment]
> please, go ahead an add tests if you feel comfortable to do so. I don't have a pressing need for this. I was just playing with reason at the time and noticed the gap.

Got it. I have 0 experience with Vim Script but I'll give it a shot as soon as I get the time :+1:

**Comment by stouf** [comment]
I updated the docs and added tests: https://github.com/stouf/ale/tree/reason-language-server
However, although I can't see any errors while the script `run-tests` is running the tests, I keep getting a "Something went wrong!" message at the end and the script terminates with an exit code `1`. I tried running the tests against the master branch and I got the same results. Below are the output of `run-tests` on both branches:

- the changes in my branch: [rls-run-tests.log](https://github.com/w0rp/ale/files/3214818/rls-run-tests.log)
- master: [master-run-tests.log](https://github.com/w0rp/ale/files/3214819/master-run-tests.log)

Is this expected or is there a problem related to my environment?


There is also another problem with reason language server itself. When actually trying ale with reason language server and opening a ReasonML file, I got an error message telling me there is a duplicate `globPattern` key in a JSON. I looked in the repository of the reason language server and thought this could be related: https://github.com/jaredly/reason-language-server/blob/6614a20364ca7ea6f0179685ee4e1cecad7ac7de/src/lsp/Main.re#L74-L75
Someone else apparently ran into the same problem and opened an issue a few hours ago: https://github.com/jaredly/reason-language-server/issues/290
I guess we can't add support for reason language server for the moment; at least until they resolve that issue 😕

@deecewan By the way, should I make a new pull request to this repository or do you want me to send a pull request to the branch this pull request is based on?

**Comment by w0rp** [comment]
The issue with the tests here is that the table of contents needs to be updated and the headings in the documentation and the table of contents need to be in alphabetical order. You can see that from the diff in the output.

**Comment by stouf** [comment]
> The issue with the tests here is that the table of contents needs to be updated and the headings in the documentation and the table of contents need to be in alphabetical order. You can see that from the diff in the output.

Right, I see them now. I'm sorry, I have no idea how I could miss them 😅 Regardless, I still get the same errors when I run the tests from the master branch 😞 Check out the diff between the two files I attached to my comment above:

```diff
13c13
< Starting Vader: 509 suite(s), 2020 case(s)
---
> Starting Vader: 510 suite(s), 2023 case(s)
1780a1781,1785
> [0m  Starting Vader: /testplugin/test/command_callback/test_reason_reason_ls_command_callbacks.vader
> [0m    (1/3) [EXECUTE] The language should be correct
> [0m    (2/3) [EXECUTE] The configured executable should be found
> [0m    (3/3) [EXECUTE] The project root should be detected correctly
> [0;32m  Success/Total: 3/3
3396,3397c3401,3402
< [0m[0;32mSuccess/Total: 2020/2020 (assertions: 2842/2842)
< [0mElapsed time: 113.304487 sec.
---
> [0m[0;32mSuccess/Total: 2023/2023 (assertions: 2846/2846)
> [0mElapsed time: 113.933437 sec.
3403c3408
< Starting Vader: 509 suite(s), 2020 case(s)
---
> Starting Vader: 510 suite(s), 2023 case(s)
5170a5176,5180
> [0m  Starting Vader: /testplugin/test/command_callback/test_reason_reason_ls_command_callbacks.vader
> [0m    (1/3) [EXECUTE] The language should be correct
> [0m    (2/3) [EXECUTE] The configured executable should be found
> [0m    (3/3) [EXECUTE] The project root should be detected correctly
> [0;32m  Success/Total: 3/3
6786,6787c6796,6797
< [0m[0;32mSuccess/Total: 2020/2020 (assertions: 2850/2850)
< [0mElapsed time: 47.358837 sec.
---
> [0m[0;32mSuccess/Total: 2023/2023 (assertions: 2854/2854)
> [0mElapsed time: 43.954581 sec.
6793c6803
< Starting Vader: 509 suite(s), 2020 case(s)
---
> Starting Vader: 510 suite(s), 2023 case(s)
8560a8571,8575
> [0m  Starting Vader: /testplugin/test/command_callback/test_reason_reason_ls_command_callbacks.vader
> [0m    (1/3) [EXECUTE] The language should be correct
> [0m    (2/3) [EXECUTE] The configured executable should be found
> [0m    (3/3) [EXECUTE] The project root should be detected correctly
> [0;32m  Success/Total: 3/3
10176,10177c10191,10192
< [0m[0;32mSuccess/Total: 2020/2020 (assertions: 2853/2853)
< [0mElapsed time: 61.291168 sec.
---
> [0m[0;32mSuccess/Total: 2023/2023 (assertions: 2857/2857)
> [0mElapsed time: 60.667099 sec.
10183c10198
< Starting Vader: 509 suite(s), 2020 case(s)
---
> Starting Vader: 510 suite(s), 2023 case(s)
11950a11966,11970
> [0m  Starting Vader: /testplugin/test/command_callback/test_reason_reason_ls_command_callbacks.vader
> [0m    (1/3) [EXECUTE] The language should be correct
> [0m    (2/3) [EXECUTE] The configured executable should be found
> [0m    (3/3) [EXECUTE] The project root should be detected correctly
> [0;32m  Success/Total: 3/3
13566,13567c13586,13587
< [0m[0;32mSuccess/Total: 2020/2020 (assertions: 2862/2862)
< [0mElapsed time: 126.849439 sec.
---
> [0m[0;32mSuccess/Total: 2023/2023 (assertions: 2866/2866)
> [0mElapsed time: 132.467553 sec.
```

Could it be related to my environment?? 🤔

**Comment by w0rp** [comment]
I have merged this and added documentation and test for it now.

Cheers! :beers:

**Comment by stouf** [comment]
Awesome, thank you !
Now let's hope they fix https://github.com/jaredly/reason-language-server/issues/290 🙏
